Updates to add new Catchment-CN version: Catchment-CN5.1

gmao-jkolassa opened the pull request:
This PR introduces the new version Catchment-CN5.1 land model, based on the Community Terrestrial System Model version 5.1 (CTSM5.1). Most of the files related to Catchment-CN5.1 are contained in the subdirectory ./GEOScatchCNCLM51_GridComp.

Documentation for Catchment-CN5.1 was added here: `[..]/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/README_CN51.md`

Source code for Catchment-CN4.5 was removed throughout, except in the make_bcs package where the generation of model parameters for Catchment-CN4.5 is still needed for Catchment-CN5.1.

It is expected that the "remap_restarts" package is zero-diff for Catchment-CN4.0.  There were, however, several bugs in CatchmentCNRst.F90 for Catchment-CN4.5 that were fixed for Catchment-CN5.1.  

**Related PRs** (in the order in which they should be merged): 
- [GEOS_Util PR 149](https://github.com/GEOS-ESM/GEOS_Util/pull/149)
- [GEOSgcm_GridComp PR 1129](https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/1129) -- this PR
- [GEOSgcm_App PR 751](https://github.com/GEOS-ESM/GEOSgcm_App/pull/751)
- [GEOSldas_GridComp PR 132](https://github.com/GEOS-ESM/GEOSldas_GridComp/pull/132)
- [GEOSldas PR 829](https://github.com/GEOS-ESM/GEOSldas/pull/829)

**Outstanding issues to be resolved before merge:**
- [ ] Add GEOSldas 0-diff test for CatchCN51.
- [x] Standard 0-diff tests for GCM. (Completed by @biljanaorescanin on 7 Aug 2025.) 
- [x] Standard 0-diff tests for GEOSldas (Completed by @gmao-rreichle after commit 3438674e2ff11d6f8b303f57cfb8867b0ad7046f.)
- [x] 0-diff test for remapping Catchment-CN4.0 restart. (Completed by @biljanaorescanin on 7 Aug 2025.)
- [x] Verify "remap_restarts" for Catchment-CN5.1.  (Completed by @biljanaorescanin on 19 Aug 2025.)
- [x] Add name of CTSM51 namelist file to GEOS_SurfaceGridComp.rc.  (Verified by @biljanaorescanin on 18 Aug 2025.)
- [x] Verify 0-diff vs. Jana/Eunjee global spin-up simulation.  (Completed by @biljanaorescanin on 3 Sep 2025.)
- [ ] Make CatchCN51 work in the GCM.  (Some of this is already in https://github.com/GEOS-ESM/GEOSgcm_App/pull/751.)
- [ ] Science verification of SIF output.
- [ ] Look into nitrogen imbalance.  Warnings were turned off temporarily (67581a99a98befe29833e797c34b51f11ff64f9c; c742c1af6efa8b11b2cb3a683e51f97753325c58).  Could be caused by error in CNPFT restart variable indexing.
- [ ] Understand why nitrogen imbalance **error** statement is printed but run does **not** abort (see https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/1152)
- [ ] Understand why T2MMIN5D values are seemingly wrong in restart file.
- [ ] Make sure calculations for T2MMIN5D are good.  See inline comments https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/1129#pullrequestreview-3187236317.

**To be addressed in a future PR:**
- Improve efficiency:
  - Move computation of cumulative states and fluxes after call to catchcn().  See https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/1129#discussion_r2193468393. 
  - Remove 90-minute sum ("m") InternalSpecs from catchcn restart file by (i) enforcing a simulation restart time and duration that is consistent with the 90-minute CN_Driver() time step and (ii) switching the order of the calls to catchment() and CN_Driver(). 



cc: @gmao-rreichle , @rdkoster , @weiyuan-jiang , @biljanaorescanin

gmao-jkolassa committed:
remove initialization of root fraction as it is initialized in CN_init

gmao-jkolassa committed:
set accumulation flag to true until we have a spun-up CNCLM51 restart file

gmao-jkolassa committed:
add cold start initialization for water flux variables

gmao-jkolassa committed:
add initialization of soil biogeochemical competition

gmao-jkolassa committed:
add missing local variables

gmao-jkolassa committed:
add missing use statements

gmao-jkolassa committed:
add missing use statements

gmao-jkolassa committed:
add allocation for missing variables

gmao-jkolassa committed:
correct allocation

gmao-jkolassa committed:
add fire method initialization

gmao-jkolassa committed:
move dt initialization for CN to before CN routines are called

gmao-jkolassa committed:
add nstep initialization

gmao-jkolassa committed:
include first_step detection needed for CLM annual routines

gmao-jkolassa committed:
change referencing of CNVeg states

gmao-jkolassa committed:
add missing use statement

gmao-jkolassa committed:
add subroutine to get date at beginning of time step

gmao-jkolassa committed:
add missing use statement

gmao-jkolassa committed:
create first time step check for CN

gmao-jkolassa committed:
add cold start initialization for some CN product variables

gmao-jkolassa committed:
add cold start initialization for more carbon flux variables

gmao-jkolassa committed:
initialize index variable

gmao-jkolassa committed:
adding cold start initializations

gmao-jkolassa committed:
fix initialization of fluxes

gmao-jkolassa committed:
add CN balance type initialization

gmao-jkolassa committed:
update balance type save

gmao-jkolassa committed:
adjusting order of initialization

gmao-jkolassa committed:
fix initializations

gmao-jkolassa committed:
activate treatment of patches

gmao-jkolassa committed:
compute column-level plant nitrogen demand from patch level demand

gmao-jkolassa committed:
initialize nitrogen demand to avoid NaN summation in column-level computation

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
initialize pacth varaibles that go through p2c as spval

gmao-jkolassa committed:
initialize pacth varaibles that go through p2c as spval

gmao-jkolassa committed:
initialize data to spval that is passed to p2g

gmao-jkolassa committed:
fix initialization of weights at patch level

gmao-jkolassa committed:
add accidentally deleted line back

gmao-jkolassa committed:
add initialization for radiation time

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
fixing variable declaration to account for compiler difference

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
correct weight calculation

gmao-jkolassa committed:
pass first flag as an input argument to CN_Driver

gmao-jkolassa committed:
code changes to account for fewer PFTs per tile in CatchCN5.1

gmao-jkolassa committed:
move getids.F90 and cleanup (PR#770)

gmao-jkolassa committed:
typo fixes

gmao-jkolassa committed:
change order of if-statements

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
update to develop

gmao-jkolassa committed:
cleanup after develop merge

gmao-jkolassa committed:
add missing variable declaration

gmao-jkolassa committed:
add missing variable declaration

gmao-jkolassa committed:
add missing variable declaration

gmao-jkolassa committed:
fixing if-statement

gmao-jkolassa committed:
hardcode CatchCN4.0/4.5 (not 5.1) option in soon to be obsolete mk_GEOSldasRestarts.F90

gmao-jkolassa committed:
hardcoding CatchCN4.0/4.5 versions in soon to be obsolete restart routines

gmao-jkolassa committed:
correct LONG_NAME for absorbed and emitted longwave flux

gmao-jkolassa committed:
changing hardcoded references to 4 PFTs to 2 PFTs

gmao-jkolassa committed:
change hardcoded references to 4 PFTs to 2

gmao-jkolassa committed:
change hardcoded references to 4 PFTs to 2

gmao-jkolassa committed:
change hardcoded references to 4 PFTs to 2

gmao-jkolassa committed:
initialize nitrogen product fluxes to zero in all patches

gmao-jkolassa committed:
initialize vcmax to cold start value of 1 in active patches

gmao-jkolassa committed:
add initialization for photosynthesis variables at every time step

gmao-jkolassa committed:
only add vegetated patches to soil patch filter

gmao-jkolassa committed:
fix restart variable size check

gmao-jkolassa committed:
remove print statement that is no longer needed

gmao-jkolassa committed:
bug fix in aggregation to gridcell level

gmao-jkolassa committed:
change calculation of soil to root conductance in first soil layer

gmao-jkolassa committed:
update to latest develop

gmao-jkolassa committed:
change long_name for CDCR2

gmao-jkolassa committed:
manually add wrapper and other missing changes

gmao-jkolassa committed:
bug fixes

gmao-jkolassa committed:
fix dtcn bug

gmao-jkolassa committed:
remove RUNSURF as export variable since it is now included in the restart file

gmao-jkolassa committed:
add psn and lmr variables back for now

gmao-jkolassa committed:
initialize all used canopy state values to suggested cold start values

gmao-jkolassa committed:
change leaf_mr_vcm parameter to one corresponding to leaf maintenance respiration method

gmao-jkolassa committed:
change cold start initialization of alphapsn

gmao-jkolassa committed:
bug fix in density calculation and setting NaN output to 0

gmao-jkolassa committed:
update to latest develop

gmao-jkolassa committed:
manually added snow albedo changes in GC

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
assigning 90-minute avg photosynthesis variables to CLM types

gmao-jkolassa committed:
fix indexing

gmao-jkolassa committed:
add phenology calculations from albedo module

gmao-jkolassa committed:
fix pointers in CN_exit and indexing of dayl

gmao-jkolassa committed:
fix typo in CN_exit call

gmao-jkolassa committed:
adjust longitude values to [0 360] convention for CLM

gmao-jkolassa committed:
update to latest develop

gmao-jkolassa committed:
keep Catchment-CN5.1 as LSM_CHOICE=4

gmao-jkolassa committed:
add missing read for mortality parameters

gmao-jkolassa committed:
replace constant water stress threshold with dependence on local wilting point

gmao-jkolassa committed:
set fcur and leaf_long parameters for non-evergreen types to values used in Catchment-CN4.0 and 4.5

gmao-jkolassa committed:
pass psn, lmr, and lai as separate variables to CLM routines instead of passing their product

gmao-jkolassa committed:
add wpwet as input to CN_Driver

gmao-jkolassa committed:
adding LAI sum fo CN as restart variable

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
add autotrophic and heterotrophic respiration as outputs

gmao-jkolassa committed:
enabling use_felxibleCN

gmao-jkolassa committed:
enable allocation of felxible_CN nutrient competition method

gmao-jkolassa committed:
provide wetness from hydrological zones mapped to carbon zones to photosynthesis routines

gmao-jkolassa committed:
change default configuration

gmao-jkolassa committed:
count for number of time steps starts at 0 to be consistent with CLM

gmao-jkolassa committed:
fix ndep assignment

gmao-jkolassa committed:
enable use of FUN model

gmao-jkolassa committed:
remove vegetation matrix functionality

gmao-jkolassa committed:
add parameter read and initialization for FUN

gmao-jkolassa committed:
add 365-day running mean of ET as input to CLM routines

gmao-jkolassa committed:
add nitrogen parameter namelist read

gmao-jkolassa committed:
read photosynthesis and canopy state parameters from namelist

gmao-jkolassa committed:
add missing source file

gmao-jkolassa committed:
add missing procedure declaration

gmao-jkolassa committed:
set fire parameters in namelist file

gmao-jkolassa committed:
overwrite default parameters with parameters from namelist file

gmao-jkolassa committed:
add missing subroutine end statement

gmao-jkolassa committed:
bug fixes

gmao-jkolassa committed:
add root profile initialization

gmao-jkolassa committed:
fix root profile namelist read

gmao-jkolassa committed:
enable CNFUN

gmao-jkolassa committed:
add initialization for several nitrogen states

gmao-jkolassa committed:
add soil nitrification/denitrification calculations

gmao-jkolassa committed:
add soil NH4 and NO3 as restart variables

gmao-jkolassa committed:
bug fixes

gmao-jkolassa committed:
read shared parameters from namelist; comment resetting of soil_decomp levels

gmao-jkolassa committed:
read additional soil parameters

gmao-jkolassa committed:
temporary sand/clay initialization

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
change to standard function for reading namelist group in SoilStateInitTimeConstMod.F90

gmao-jkolassa committed:
remove unnecessary variables

gmao-jkolassa committed:
bug fixes

gmao-jkolassa committed:
update time step counter

gmao-jkolassa committed:
adding inclusion of get_nstep in CN_Driver

gmao-jkolassa committed:
bug fix in root respiration calculation

gmao-jkolassa committed:
add heterotrophic respiration from coarse woody debris as on of the heterotrophic respiration components

gmao-jkolassa committed:
bug fixes in configuration

gmao-jkolassa committed:
fix nlevdecomp_full definition

gmao-jkolassa committed:
increase istep_cn

gmao-jkolassa committed:
merge develop

gmao-jkolassa committed:
Merge branch 'develop' into feature/jkolassa_cnclm51

weiyuan-jiang committed:
fix N_CONSTIT and get_nstep kind

weiyuan-jiang committed:
Merge branch 'develop' into feature/jkolassa_cnclm51

gmao-jkolassa committed:
manual update to align with latest develop branch

gmao-jkolassa committed:
move initialization of CNFUN variables to before CNDriver call

gmao-jkolassa committed:
fix permissions test

gmao-jkolassa committed:
fix permissions

gmao-jkolassa committed:
fix permission

gmao-jkolassa committed:
use consistent naming convention

gmao-jkolassa committed:
Merge branch 'develop' into feature/jkolassa_cnclm51

gmao-jkolassa committed:
change default settings for accumulation flags

gmao-jkolassa committed:
update CN_init module name to be consistent with naming convention

gmao-rreichle committed:
updated syntax to ESMF::ESMF (GEOScatchCNCLM51_GridComp/CMakeLists.txt)

gmao-rreichle committed:
clarified PRESCRIBE_DVG implementation: only applies to CatchCNCLM40 (GEOS_SurfaceGridComp.rc)

gmao-rreichle committed:
clarified choice of Z0_FORMULATION, applies to Catch only (GEOS_SurfaceGridComp.rc)

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Shared/GEOS_SurfaceGridComp.rc:
```diff
@@ -69,6 +69,8 @@
 
 # ---- Formulation for turbulent roughness length (z0)
 #
+#    Choice implemented for Catch only.  Hardwired in CatchCN.
```
Does the "Z0_FORMULATION" that is hardwired for CN40 and CN51 correspond to any of the choices for Catch?

gmao-jkolassa replied:
Updated Z0-formulation in GEOS_CatchCNCLM51GridComp.F90 to match the options available for Catchment. In Catchment-CN5.1, Z0 is computed using two vegetation types that correspond to the 'old' 6 types.

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/GEOS_CatchCNCLM51GridComp.F90:
```diff
@@ -0,0 +1,8682 @@
+  VERIFY_(STATUS)
+
+  call MAPL_AddInternalSpec(GC                  ,&
+    LONG_NAME          = 'ground_temperature'        ,&
```
What exactly is "ground_temperature"?  Is it the temp of the top 5 cm of the soil?  Needs better long_name

gmao-rreichle reviewed (commented):
@gmao-jkolassa : I made a first pass through the PR and added a bunch of inline comments for discussion, see below.

gmao-jkolassa commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@ esma_set_this ()
 
 add_subdirectory (GEOScatchCNCLM40_GridComp)
 add_subdirectory (GEOScatchCNCLM45_GridComp)
+add_subdirectory (GEOScatchCNCLM51_GridComp)
```
See above.

gmao-jkolassa commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/CLM51/ActiveLayerMod.F90:
```diff
@@ -0,0 +1,203 @@
+module ActiveLayerMod
```
The files are not perfect carbon copies. They have been changed where we initialize CLM variables from our restart files or - in rare cases - where we have changed parameter values or calculations from what CLM implements. Also, several of their 'MAPL-like' subroutines have been used to actually use MAPL in order to not duplicate routines. I am working on a Catchment-CN5.1 documentation that will detail all of this. I am happy to include that document in the form of a README file. I'll make sure that the exact CTSM version that was used here is stated explicitly in the documentation.

gmao-jkolassa replied:
I have included the CTSM original files that were used in developing Catchment-CN5.1 into the /CLM51/CTSM_originals directory. The Catchment-CN5.1 documentation specifies how each file has been modified and includes the name and link to the CTSM tag that was used as the starting point.

gmao-jkolassa commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/GEOS_CatchCNCLM51GridComp.F90:
```diff
@@ -0,0 +1,8682 @@
+    ! NOTE: In CNPhenologyMod.F90, init_gdd20 is always set to .false. as well. For GEOS-5 runs, need to discard at least the first 2 years.
+    ! This is not a problem for offline runs because we always spin up the model whenever we change meterology. fzeng, July 2017 
+    ! -------------------------------------------------------------------------------------------------------------------------------------- 
+    logical, parameter :: init_accum = .false.! jkolassa May 2023: needs to be set to true if no CNCLM51 restart is available
+    logical, parameter :: init_accum_365 = .false.! jkolassa May 2023: needs to be set to true if no CNCLM51 restart is available
```
> Does the comment still apply? If so, how would a user without a CNCLM51 restart know what to do? This is very deep in the code. Are we requiring somewhere that a CNCLM51 restart must be supplied?

A user should **NEVER** try to set up an experiment without a CNCLM51 restart file and should always use a fully spun-up CNCLM51 restart. The default values of these toggles are set such that the existence of a CNCLM51 restart file is assumed. These toggles really only exist for the Catchment-CN developer to be able to start a CNCLM51 from a different model restart file when starting a spin-up run (that is why I have not moved them higher up in the code). It may make sense to include a warning (or even error) in ldas_setup if a user tries to set up a CNCLM51 experiment without a CNCLM51 restart file. (Really this should be the case for any CNCLMXX experiment)

gmao-jkolassa committed:
adding CN_CLM51 namelist file to source code

gmao-jkolassa committed:
cleanup

gmao-jkolassa commented:
> The nml file should be added into the src code, probably best in the ./GEOScatchCNCLM51_GridComp/ directory. The CMakeLists.txt should be edited so the build copies this file into the install/etc directory, and CNCLM_init_mod.F90 can pick it up from there. If we can identify a few parameters that users may want to change, we can add them into GEOS_SurfaceGridComp.rc. But I don't think it makes sense to add everything that's currently in CatchCN5.1.nml into GEOS_SurfaceGridComp.rc

I have made commits to implement the suggested handling of the namelist file. The file now 'lives' in the ./GEOScatchCNCLM51_GridComp/ directory and is moved to ./install/etc when the model is built. I have made changes to the GEOSldas_GridComp to then move the file to the /run directory when ldas_setup is run. The CNCLM_init_mod.F90 module then grabs it from there. I have not implemented a check for the existence of the file in CNCLM_init_mod.F90, as such checks are already implemented in the various initialization subroutines where the file is used.

> 
> I guess we'll need a custom file for CatchCN5.1. We'll need to edit make_bcs so the (resolution-specific) file is created as needed. Within the bcs, the file should end up in ./land/[BCS_RESOLUTION]/clsm/ Are the values in this file part of the CatchCN5.1 internal restart file? We may also need to modify GEOS_Util's remap_restarts package and ldas_setup (and eventually gcm_setup).

Upon reviewing the code I realized that I have already implemented code changes to address this issue. CatchmentCNRst.F90 reads the original CLM_veg_typs_fracs file which has 19 PFTs with up to 4 PFTs per tile and implements a mapping to the new 15 Catchment-CN5.1 PFTs with 2 PFTs per tile. So no separate CLM_veg_typs_fracs is needed for Catchment-CN5.1. Apologies for the confusion, I forgot that I had already addressed this!

gmao-jkolassa committed:
update FLWALPHA value for Catchment-CN5.1 to be consistent with Catchment

gmao-jkolassa committed:
restore files to develop version, as they will become obsolete

gmao-jkolassa committed:
introduce multiple Z0-formulation options for Catchment-CN5.1

gmao-jkolassa committed:
Revert "restore files to develop version, as they will become obsolete"

This reverts commit b4073dfe5d64ce8961f1af7be3ff72a090c71da3.

gmao-jkolassa commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/mk_GEOSldasRestarts.F90:
```diff
@@ -51,8 +51,11 @@ PROGRAM mk_GEOSldasRestarts
   integer, parameter :: VAR_PFT_CLM40 = 74 ! number of CN PFT variables per column
   integer, parameter :: npft    = 19  
   integer, parameter :: npft_clm45    = 19
+  integer, parameter :: npft_clm51    = 15
```
Holding off on implementing this, because it causes issues when trying to build the model.

gmao-rreichle committed:
first pass of updating long_name attributes to match M21C file specs; minimal white-space changes (GEOS_CatchCNCLM51GridComp.F90)

gmao-jkolassa committed:
correct internal state name

gmao-jkolassa committed:
add Z0-formulation choice in Driver

gmao-rreichle committed:
minimal white-space changes for better vertical alignment (GEOS_CatchCNGridComp.F90)

gmao-rreichle committed:
removed obsolete variables and calls to MAPL_GetResource() (GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
added missing call to diagnostic snow temp routine prior to calling snow albedo routine (GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
removed stale irrigation code (GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
fixed bug in earlier commit (GEOS_CatchCNCLM51GridComp.F90)

gmao-jkolassa committed:
bug fix to Z0-formulation implementation

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
fixes to catchcn_internal statements in SetServices

gmao-rreichle committed:
Updated Z0_FORMULATION help text (GEOS_SurfaceGridComp.rc)

gmao-jkolassa committed:
remove unused declarations in SetServices

gmao-jkolassa committed:
reorganize CN_CLM51 modules

gmao-jkolassa committed:
Revert "reorganize CN_CLM51 modules"

This reverts commit b8bd86b190e21c5631e6a589fcc9fb55c91cc3df.

gmao-rreichle committed:
added land constants as HISTORY export variables (GEOS_CatchCNCLM51GridComp.F90, catchmentCN.F90)

gmao-rreichle committed:
fix land water and energy balance terms in CatchCNCLM51 (GEOS_CatchCNCLM51GridComp.F90, catchmentCN.F90)

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/GEOS_CatchCNCLM51GridComp.F90:
```diff
@@ -0,0 +1,8827 @@
+  VERIFY_(STATUS)
+
+
+  call MAPL_AddInternalSpec(GC,                    &
+       LONG_NAME          = 'overland_runoff_including_throughflow'  ,&
+       UNITS              = 'kg m-2 s-1'                ,&
+       SHORT_NAME         = 'RUNSURF'                   ,&
+       FRIENDLYTO         = trim(COMP_NAME)             ,&
+       DIMS               = MAPL_DimsTileOnly           ,&
+       VLOCATION          = MAPL_VLocationNone          ,&
+       RESTART            = MAPL_RestartOptional        ,&
+                                           RC=STATUS  )
```
Do we really need RUNSURF to be an "internal spec".  It looks like besides serving as the usual export spec for surface runoff, RUNSURF only contributes to the running mean RUNSRFM.   Why not compute cumulative sum *after* the call to catchcn(), when we have the diagnostic surface runoff from catchcn()?  At that time, we also have other diagnostics (incl. baseflow) from catchcn(), so they don't have to be computed *before* the call to catchcn().  Since these are multi-day sums/averages, it's hard to imagine that we change any science by shifting the accumulation calc by one (heartbeat) time step. (The cumulative sums/averages *are* restart variables [internal specs], so the inputs to CN_Driver() are available.) The only hiccup might be the very first time step after a "cold" restart, but we don't allow for users to do "cold" restarts anyway.

gmao-rreichle committed:
adapt GEOS_CatchCNCLM40GridComp.F90, GEOS_CatchCNCLM45GridComp.F90 to interface change of catchcn() in recent commit to fix balance terms in CatchCNCLM51

gmao-rreichle committed:
simplified if conditions for GOSWIM to be consistent with GEOS_CatchGridComp.F90 (GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
fixed FRZR in CatchCN51; informative error message for neg precip (GEOS_CatchCNCLM51GridComp.F90, GEOS_CatchGridComp.F90)

gmao-rreichle committed:
fixed permissions for CatchCN files

gmao-rreichle committed:
clarified [radians] units of lat/lon coords (GEOS_CatchCNCLM51GridComp.F90)

biljanaorescanin committed:
add to bcs new file

gmao-rreichle committed:
fixed variable name CATCHCN_INTERNAL (GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
Numerical derivatives for Monin-Obukhov exchange coefficients over land in offline mode for CatchCN51 (GEOS_CatchCNCLM51GridComp.F90, GEOS_SurfaceGridComp.rc)

gmao-jkolassa committed:
adding CTSM5.1 source code for reference; code is not built

gmao-rreichle committed:
first pass at removing CatchCNCLM45 source code

gmao-jkolassa committed:
revert obsolete mk_restarts files to their develop versions

gmao-jkolassa committed:
change long name of TG to 'surface layer soil temperature' and change long names of associated variables

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/GEOS_CatchCNCLM51GridComp.F90:
```diff
@@ -0,0 +1,8682 @@
+!  $Id$ 
```
I implemented all of the above.  As of https://github.com/GEOS-ESM/GEOSgcm_GridComp/pull/1129/commits/5dad83de0161ddf5d963e7760d7872fad7f0c45c, the code builds successfully.  However, the added functionality (such as the numerical derivatives for Monin-Obhukov exchange coeffs and the output of the land constants file) has not been verified.

biljanaorescanin committed:
change copy to list

gmao-rreichle committed:
Add CatchCNCLM51 parameter file to BCS dir

Add new file to BCS dir

gmao-jkolassa committed:
Merge branch 'develop' into feature/jkolassa_cnclm51

gmao-jkolassa committed:
pass parameter file from GC to initialization routines and remove hardcoded overwriting of parameters

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/GEOS_CatchCNCLM51GridComp.F90:
```diff
@@ -4177,11 +4379,26 @@ subroutine RUN1 ( GC, IMPORT, EXPORT, CLOCK, RC )
+! read CNCLM51 parameter file
+!-----------------------------
+
+   call MAPL_GetResource (MAPL, cnclm51_paramfile, label = 'ctsm51_params.c210923_forCNCLM_FILE:', default = 'ctsm51_params.c210923_forCNCLM.nc', RC=STATUS )
```
How about a more generic name for the rc variable?  I suggest:

```suggestion
   call MAPL_GetResource (MAPL, cnclm51_paramfile, label = 'CNCLM51_PARAM_FILE:', default = 'ctsm51_params.c210923_forCNCLM.nc', RC=STATUS )
```

We could add this rc variable (and its default) into GEOS_SurfaceGridComp.rc, so a user has an easier time locating the parameter file, if that's desirable

gmao-jkolassa replied:
The rc variable name has been changed, but it has not yet been added into the GEOS_SurfaceGridComp.rc

gmao-jkolassa committed:
adding parameter file to pftcon initialization call

gmao-jkolassa committed:
rename rc variable for CNCLM51 parameter file

gmao-jkolassa committed:
add first draft of solar induced fluorescence calculation adapted for CNCLM51

gmao-jkolassa committed:
typo fix

gmao-jkolassa committed:
bug fixes

gmao-jkolassa committed:
include originals of CTMS files that were used in Catchment-CN5.1

gmao-jkolassa committed:
change long name of CNHR and CNAR

gmao-jkolassa committed:
added disclaimer that dbg routines in CatchCN /utils are specific to Catchmnet-CN4.0 only

gmao-jkolassa commented:
> I assume this refers to the file /discover/nobackup/jkolassa/CLM/parameter_files/ctsm51_params.c210923.nc. It looks like this is a small file that is independent of the tile space (resolution of the model simulation). The file should be placed into the shared bcs directory under ./make_bcs_inputs/land/veg Then make_bcs must be modified so it makes a copy of the file into [BCS_VERSION]/land/shared/ Next, ldas_setup (and gcm_setup, and eventually fvsetup) must link the file from its BCS_VERSION location into the "input" dir (where the LAI and greenness files are read by the model executable), and CNCLM_init_mod.F90 should pick it up from there. In other words, exactly what's happening with CO2_MonthlyMean_DiurnalCycle.nc4

The parameter file is now part of the boundary conditions and is moved to the /input directory when setting up an experiment. The _CN_init()_ subroutine grabs it from there.

> The nml file should be added into the src code, probably best in the ./GEOScatchCNCLM51_GridComp/ directory. The CMakeLists.txt should be edited so the build copies this file into the install/etc directory, and CNCLM_init_mod.F90 can pick it up from there. If we can identify a few parameters that users may want to change, we can add them into GEOS_SurfaceGridComp.rc. But I don't think it makes sense to add everything that's currently in CatchCN5.1.nml into GEOS_SurfaceGridComp.rc

The namelist file is now included with the source code in the ./GEOScatchCNCLM51_GridComp/ directory and is moved to the /run directory when an experiment is set up.

> I guess we'll need a custom file for CatchCN5.1. We'll need to edit make_bcs so the (resolution-specific) file is created as needed. Within the bcs, the file should end up in ./land/[BCS_RESOLUTION]/clsm/ Are the values in this file part of the CatchCN5.1 internal restart file? We may also need to modify GEOS_Util's remap_restarts package and ldas_setup (and eventually gcm_setup).
> 

The mapping of the 19 PFTs with 4 PFTs per tile to 15 PFTs with 2 PFTs per tile is now handled by CatchmentCNRst.F90, so that all versions of Catchment-CN can use the same CLM_veg_typs_fracs file.

gmao-jkolassa commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/Shared/catchmentCN.F90:
```diff
@@ -1100,7 +1100,7 @@ SUBROUTINE CATCHCN (                                           &
           SHFLUX(N)=SHFLUX(N)+ECORR(N)*ALHE
           EVAP(N)=EVAP(N)-ECORR(N)
-          HLATN(N)=ESOI(N)*ALHE + EVEG(N)*ALHE + EINT(N)*ALHE + ESNO(N)*ALHS
+          LHFLUX(N)=ESOI(N)*ALHE + EVEG(N)*ALHE + EINT(N)*ALHE + ESNO(N)*ALHS
```
Following Randy's investigation, both models should remain as they are. The differences are due to the following (from Randy):

> "in the original Catchment.F90, we separated the bare soil evaporation from the transpiration using some resistance calculations, as made possible from the use of a single surface temperature to represent both the canopy and the bare soil.  In CatchCN, such an approach was no longer possible given that we follow separately the ground temperature and the canopy temperature, with the former being relevant to bare soil evaporation and the latter to transpiration.  That is, the code differences you show are a simple reflection of the need for a different approach in CatchCN."

gmao-rreichle replied:
If I understand Randy's answer correctly, the block in CatchCN that "UPDATE[S] SENSIBLE HEAT IF WATER LIMITATIONS WERE IMPOSED" does not apply to Catch, so no action is needed for this block. 
But I'm not sure if Randy's answer addresses the correction for "EXCESSIVE INTERCEPTION LOSS" in Catchment just prior to WUPDAT().  This seems related to the (water holding) capacity of the interception reservoir, which shouldn't be related to the treatment of soil/veg temperatures.  Do we want to include this correction in CatchCN?

gmao-rreichle replied:
Follow-up comment from Randy:  
"_In Catchment, EIRFRC is the diagnosed fraction of the total ET that is made up of interception loss.  The diagnosis calculations are based on the simple resistance network that Catchment uses, one wherein the surface temperature and canopy temperature are equal.  This is not possible with CatchCN, which uses separate canopy and surface temperatures.  In fact, EIRFRC does not exist in CatchCN; the more complex resistance network means that we don't have the luxury of diagnosing it simply.  The possibility that the computed interception loss exceeds the canopy reservoir amount is dealt with instead in WUPDAT, with a new variable ECORR (not necessary for Catchment) that is used after the fact to adjust the sensible and latent heat fluxes accordingly._"

I added brief comments into catchment.F90 and catchmentCN.F90 stating that the Catchment/CatchmentCN differences are intentional.

weiyuan-jiang committed:
bug fix

gmao-jkolassa commented:
> @gmao-jkolassa
> 
> My last commit has some fixes that make remap_restart work. please take a look and make sure it is right. Some variables are read in without conditions
> 
> https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/7b27aa958e5efbd7b8e8bc16c60377035b5485ea/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90#L227-L230
> 
> But when they are written out, there are conditions
> 
> https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/7b27aa958e5efbd7b8e8bc16c60377035b5485ea/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90#L364-L367
> 
> I felt it is inconsistent.

Thanks @weiyuan-jiang for taking the time to look through this and make the necessary fixes!

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90:
```diff
@@ -278,6 +336,36 @@ subroutine write_nc4(this, filename, rc)
+
+     elseif (this%isCLM51) then
+
+         call MAPL_VarWrite(formatter,"SFMM",  this%sfmm)
+
+          call MAPL_VarWrite(formatter,"ABM",     this%ABM, rc =rc     )
+          call MAPL_VarWrite(formatter,"FIELDCAP",this%FIELDCAP)
+          call MAPL_VarWrite(formatter,"HDM",     this%HDM     )
+          call MAPL_VarWrite(formatter,"GDP",     this%GDP     )
+          call MAPL_VarWrite(formatter,"PEATF",   this%PEATF   )
+          call MAPL_VarWrite(formatter,"RHM",     this%RHM)
+          call MAPL_VarWrite(formatter,"WINDM",   this%WINDM)
+          call MAPL_VarWrite(formatter,"RAINFM",  this%RAINFM)
+          call MAPL_VarWrite(formatter,"SNOWFM",  this%SNOWFM)
+          call MAPL_VarWrite(formatter,"RUNSRFM", this%RUNSRFM)
+          call MAPL_VarWrite(formatter,"AR1M",    this%AR1M)
+          call MAPL_VarWrite(formatter,"SNDZM5D", this%SNDZM5D)
+          call MAPL_VarWrite(formatter,"T2M10D",  this%T2M10D)
+          call MAPL_VarWrite(formatter,"T2MMIN5D",this%T2MMIN5D)
+          call MAPL_VarWrite(formatter,"TG10D",   this%TG10D)
+          call MAPL_VarWrite(formatter,"RH30D",   this%RH30D)
+          call MAPL_VarWrite(formatter,"TPREC10D",this%TPREC10D)
+          call MAPL_VarWrite(formatter,"TPREC60D",this%TPREC60D)
+          call MAPL_VarWrite(formatter,"LMRSUNM", this%LMRSUNM )
+          call MAPL_VarWrite(formatter,"LMRSHAM", this%LMRSHAM )
+          call MAPL_VarWrite(formatter,"LAISUNM", this%LAISUNM )
+          call MAPL_VarWrite(formatter,"LAISHAM", this%LAISHAM )
+
```
See comment above about changing the if blocks for these statements.
Also note Weiyuan's comment about having consistency across the VarRead and VarWrite blocks.

gmao-rreichle reviewed (commented):
See comments below:

gmao-rreichle committed:
moved CLM51/CTSM_original to CLM51_orig_files/, fixed file permissions

gmao-rreichle committed:
added comments about a specific Catchment vs CatchCN difference (catchment.F90, catchmentCN.F90)

gmao-rreichle committed:
move update_model_para4cn.F90 out of ./CLM51 directory

gmao-rreichle committed:
removed vim swap files from repository; added orig version of CLM51 file cmake/genf90_utils.cmake

gmao-rreichle committed:
moving update_model_para4cn.F90 back into ./CLM51/

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/CLM51/update_model_para4cn.F90:
```diff
@@ -5,8 +5,11 @@ MODULE update_model_para4cn
+       prev_year,prev_month,prev_day,prev_dofyr,prev_hour,prev_min,prev_sec, &
+       LocalTileID
 
   public :: upd_curr_date_time, upd_tileid
```
Do we still need subroutine `upd_tileid()`?  I could not find it used anywhere in this repo.  (Perhaps it was used with CatchCNCLM45, which has been removed.)

gmao-rreichle reviewed (commented):
A couple more comments below:

gmao-rreichle committed:
fixed/cleaned up InternalSpecs for CNCLM40 and CNCLM51 (CatchmentCNRst.F90):
- reordered lists of InternalSpecs to match that of MAPL_AddInternalSpec() calls in GEOS_CatchCNCLM*GridComp.F90
  - type declarations
  - MAPL_VarRead
  - MAPL_VarWrite
  - allocates
- added missing InternalSpecs for CNCLM51: sfmm, et365d, runsurf
- renamed field t2 -> cli_t2m for consistency with name of InternalSpec
- clean-up (vertical alignment, white space, ...)

gmao-rreichle committed:
removed CNCLM45 from mk_restarts (CatchmentCNRst.F90, clm_varpar_shared.F90)

gmao-rreichle committed:
additional fixes and cleanup of InternalSpecs (CatchmentCNRst.F90):
- added missing (?) variables
- reordered variables for consistency
- added comments
- fixed indent
- improved vertical alignment

gmao-rreichle committed:
fixed bug in previous commit (CatchmentCNRst.F90)

gmao-rreichle committed:
bug fix: added missing variable declaration (CatchmentCNRst.F90)

gmao-rreichle committed:
added comment to TSURF InternalSpec (GEOS_CatchGridComp*.F90)

gmao-rreichle committed:
cleaned up CatchCN dimension variable names

gmao-rreichle committed:
fix bug in previous commit (CLM??/clm_varpar.F90)

gmao-jkolassa committed:
add first draft of Catchment-CN5.1 README file

gmao-rreichle committed:
bug fix: 2nd dimension of TG; cleaned up temporary comments (CatchmentCNRst.F90)

gmao-jkolassa committed:
update Catchment-CN5.1 README file

gmao-jkolassa committed:
add original shr_log_mod.F90 to CLM51_orig_files

gmao-rreichle committed:
edited comments in makebcs tools to clarify parameters for CatchmentCN versions (mkCatchParam.F90, mod_process_hres_data.F90)

gmao-jkolassa committed:
fix TG allocation dimensions

gmao-rreichle committed:
improved vertical alignment and indentation (CLM51/clm_varpar.F90)

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90:
```diff
@@ -156,6 +181,28 @@ function CatchmentCNRst_create(filename, cnclm, time, rc) result (catch)
+           call MAPL_VarRead(formatter,"RH30D",   catch%RH30D   , __RC__)
+           call MAPL_VarRead(formatter,"TPREC10D",catch%TPREC10D, __RC__)
+           call MAPL_VarRead(formatter,"TPREC60D",catch%TPREC60D, __RC__)
+        endif
```
I cleaned up CatchmentCNRst.F90.

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90:
```diff
@@ -774,561 +962,677 @@ subroutine re_tile(this, InTileFile, OutBcsDir, OutTileFile, surflay, rc)
+                     var_pft_out(n, nz,nv,70) = max(var_pft_out(n, nz,nv,70),0.)
+                     var_pft_out(n, nz,nv,73) = max(var_pft_out(n, nz,nv,73),0.)
+                     var_pft_out(n, nz,nv,74) = max(var_pft_out(n, nz,nv,74),0.)
+                     !if(this%isCLM45) var_pft_out(n, nz,nv,75) = max(var_pft_out(n, nz,nv,75),0.)
+                     if(this%isCLM51) then
```
@gmao-jkolassa : For CNCLM51, element 75 isn't populated.  Is this ok?    Also, there is a big block of comments starting about 30 lines down that outlines array indices for CNCLM40 and CNCLM45, but there is no mention of CNCLM51.   Are these comments accurate for CNCLM40 and CNCLM45?  Can we translate the CNCLM45 into CNCLM51 comments, so they are useful now that CNCLM45 has been removed?

gmao-jkolassa replied:
I replaced the CNCLM45 variable description with the for CNCLM5.1 and in the process also cleaned up the variable description for CNCLM40, which had some obsolete information.

gmao-rreichle committed:
improved vertical alignment (white-space changes only) (CatchmentCNRst.F90)

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90:
```diff
@@ -774,561 +962,677 @@ subroutine re_tile(this, InTileFile, OutBcsDir, OutTileFile, surflay, rc)
+          !                      HDM(:)            ! not done (parameter)
+          !                 FIELDCAP(:)            ! not done (parameter)
+          
+          var_out = this%rhm     (this%id_glb(:));  this%rhm     = var_out
```
@gmao-jkolassa : In reorganizing these statements, I fixed that seemed like a bug.  Up until f0acab1e3a99a229b3c228317afb1b1630c254c0 (1 Aug 2025), the statements here for ar1m, rainfall, rhm, and a few others were within a "isCLM45" (only) block.   Please advise if you think the original code did not have a bug.

gmao-jkolassa replied:
The updated code looks correct to me.

gmao-rreichle reviewed (commented):
See comments below

weiyuan-jiang commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90:
```diff
@@ -774,561 +962,677 @@ subroutine re_tile(this, InTileFile, OutBcsDir, OutTileFile, surflay, rc)
-           this%sfmm = var_out_zone
-        endif
+       where(isnan(var_off_pft))          var_off_pft = 0.
+       where(var_off_pft /= var_off_pft)  var_off_pft = 0.     ! TMPRR: does this do anything? 
```
> @gmao-jkolassa @weiyuan-jiang : This statement seems odd. Maybe the inequality turns to true if there are no-data-values? If so, we should add a comment, something like "replace NaNs with 0"

I think  "Nan /= Nan" is true. So both where statements do the same thing.

gmao-jkolassa committed:
temporarily disable nwarning printout to avoid very large log files; the nwarning issue needs to be looked into if it persists after spinup is complete

gmao-jkolassa commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90:
```diff
@@ -774,561 +962,677 @@ subroutine re_tile(this, InTileFile, OutBcsDir, OutTileFile, surflay, rc)
+       i = 1
+       do jj = 1,VAR_COL
+          do nz = 1,nzone
+             var_off_col(:,nz,jj) = this%cncol(:,i)            ! TMPRR: will be inverted again in regrid_carbon() below
```
Again, this predates me, so I can only guess. But it seems that this may be done to facilitate the indexing over different variables for some of the calculations in _regrid_carbon()_. For example to make it easier to do checks like this
`if(var_col_out (n, nz,14) < 10.) then`

gmao-jkolassa committed:
add missing element for Catchment-CN5.1

gmao-rreichle committed:
cleaned up comments (CatchmentCNRst.F90)

gmao-rreichle committed:
commented out unused subroutine upd_tileid()   (CLM51/update_model_para4cn.F90)

gmao-jkolassa committed:
update PFT type and fraction assignment for Catchment-CN5.1

gmao-jkolassa commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90:
```diff
@@ -774,561 +962,677 @@ subroutine re_tile(this, InTileFile, OutBcsDir, OutTileFile, surflay, rc)
+         CLMC_pf1 = this%fvg(  :,1)
+         CLMC_pf2 = this%fvg(  :,2)
+         CLMC_sf1 = this%fvg(  :,3)
+         CLMC_sf2 = this%fvg(  :,4)
```
I pushed a commit to update this block. Because Catchment-CN5.1 only has 2 PFTs per tile, only CLMC_pt1, CLMC_st1, CLMC_pf1, and CLMC_sf1 need to be assigned.

gmao-jkolassa committed:
replace description of CLM45 variables wih CTSM5.1 variables; cleanup of CLM40 variables description

gmao-rreichle committed:
consistent use of "use shr_kind_mod" in CLM51/*.F90

gmao-rreichle committed:
fixed syntax error in previous commit (shr_kind_mod.F90)

gmao-rreichle committed:
Add name of CNCLM51 nml file to GEOS_SurfaceGridComp.rc (catch_wrap_state.F90, GEOS_CatchCNGridComp.F90, CNCLM_init_mod.F90)

gmao-rreichle committed:
fix previous commit (GEOS_CatchCNCLM51GridComp.F90, CLM51/CNCLM_init_mod.F90)

biljanaorescanin committed:
initializing the component’s internal config flag

gmao-rreichle committed:
edited comment; white-space changes (GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
fixed treatment of rc variables in SetServices() (GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
edited/added comments about handling of rc variables (GEOSland_GridComp/GEOScatchCN_GridComp/GEOS_CatchCNGridComp.F90)

gmao-rreichle committed:
edited/added comments about handling of rc variables (GEOS_CatchGridComp.F90)

gmao-rreichle committed:
revised handling of rc parameters for CatchCN (GEOS_CatchCNGridComp.F90)

biljanaorescanin committed:
Merge pull request #1142 from GEOS-ESM/borescan/catchCN_clm_5.1_fix

gmao-rreichle committed:
some clean-up of implementation of Z0_FORMULATION; disabled outdated and bad Z0_FORMULATION=1 (GEOS_SurfaceGridComp.rc, lsm_routines.F90, GEOS_CatchGridComp.F90, GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
fix "include" and "use" errors in previous commit (lsm_routines.F90, GEOS_CatchGridComp.F90, GEOS_CatchCNCLM51GridComp.F90)

gmao-rreichle committed:
fixed error in use of _ASSERT() (lsm_routines.F90)

gmao-rreichle committed:
Merge branch 'develop' into feature/jkolassa_cnclm51

gmao-rreichle committed:
fix GNU build error (lsm_routines.F90)

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/Shared/lsm_routines.F90:
```diff
@@ -2729,6 +2717,54 @@ subroutine dampen_tc_oscillations( dtstep, tair, tc_old, tc_new_in, &
+
+    case default
+
+       _ASSERT(.FALSE., 'unknown Z0_FORMULATION')
```
> GNU fails to recognize that is a comment

Interesting!  I changed _ASSERT() to ASSERT().  With any luck, it'll build with GNU now.  Thanks!

gmao-rreichle committed:
commented out "nbalance warning at g" print statements (CNBalanceCheckMod.F90)

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/GEOS_CatchCNCLM51GridComp.F90:
```diff
@@ -6631,23 +7133,65 @@ subroutine Driver ( RC )
+
+         do n = 1,ntiles
+            ta_count(n) = ta_count(n) + 1
+            TA_MIN(n) = min(TA_MIN(n),TA(n))
```
The minimum air temperature used here is the "lowest-model-level" temperature, which is typically at about 40-60 m height, and not the screen-level (2m) temperature.   There is a comment above (just before the if block) that suggests as much, but it's still misleading to label this variable T2m and claim as much in the LONG_NAME of the Internal_Spec definition.

gmao-rreichle commented on GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/GEOSland_GridComp/GEOScatchCN_GridComp/GEOScatchCNCLM51_GridComp/GEOS_CatchCNCLM51GridComp.F90:
```diff
@@ -6631,23 +7133,65 @@ subroutine Driver ( RC )
+            TA_MIN(n) = min(TA_MIN(n),TA(n))
+
+            if (ta_count(n) == n1d) then
+                T2MMIN5D(n)   = ((accper-1)*T2MMIN5D(n)   + TA_MIN(n))  / accper
```
Since init_accum is hardwired to .false., "accper" here is undefined.  This seems to be a bug.

gmao-rreichle reviewed (commented):
The treatment of T2MMIN5D doesn't seem right.  See comments below

gmao-rreichle committed:
fix typo in 4 Aug 2025 bug fix of remap_restart utility for CatchCN51 (CatchmentCNRst.F90)

gmao-rreichle committed:
clarify use of CN time step counter (istep) (GEOS_CatchCNCLM51GridComp.F90, CNCLM_DriverMod.F90)

gmao-rreichle committed:
corrected LONG_NAME attributes, comments, units for multi-day exponential moving averages (GEOS_CatchCNCLM51GridComp.F90)

biljanaorescanin committed:
Merge branch 'develop' into feature/jkolassa_cnclm51

gmao-rreichle committed:
fixed indent, improved vertical alignment, edited comments, removed redundant "implicit none" (CNCLM_DriverMod.F90)
